server/base/src/main/java/org/apache/accumulo/server/util/FindCompactionTmpFiles.java

```diff
@@ -0,0 +1,205 @@
+      throws InterruptedException {
+    String tablePattern = tableId != null ? tableId : "*";
+    final String pattern = "/tables/" + tablePattern + "/*/*";
+    final Collection<Volume> vols = context.getVolumeManager().getVolumes();
+    final ExecutorService svc = Executors.newFixedThreadPool(vols.size());
+    final List<Path> matches = new ArrayList<>(1024);
+    final List<Future<Void>> futures = new ArrayList<>(vols.size());
+    for (Volume vol : vols) {
+      final Path volPattern = new Path(vol.getBasePath() + pattern);
+      LOG.info("Looking for tmp files that match pattern: {}", volPattern);
+      futures.add(svc.submit(() -> {
+        try {
+          FileStatus[] files = vol.getFileSystem().globStatus(volPattern,
+              (p) -> p.getName().contains("_tmp_" + ExternalCompactionId.PREFIX));
+          Arrays.stream(files).forEach(fs -> matches.add(fs.getPath()));
+        } catch (IOException e) {
+          LOG.error("Error looking for tmp files in volume: {}", vol, e);
+        }
+        return null;
+      }));
+    }
+    svc.shutdown();
+
+    while (futures.size() > 0) {
+      UtilWaitThread.sleep(10_000);
+      Iterator<Future<Void>> iter = futures.iterator();
+      while (iter.hasNext()) {
+        Future<Void> future = iter.next();
+        if (future.isDone()) {
+          iter.remove();
+          try {
+            future.get();
+          } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException("Error getting list of tmp files", e);
+          }
+        }
+      }
+    }
+    svc.awaitTermination(10, TimeUnit.MINUTES);
+    LOG.debug("Found compaction tmp files: {}", matches);
+
+    // Remove paths of all active external compaction output files from the set of
+    // tmp files found on the filesystem. This must be done *after* gathering the
+    // matches on the filesystem.
+    for (DataLevel level : DataLevel.values()) {
+      context.getAmple().readTablets().forLevel(level).fetch(ColumnType.ECOMP).build()
+          .forEach(tm -> {
+            tm.getExternalCompactions().values()
+                .forEach(ecm -> matches.remove(ecm.getCompactTmpName().getPath()));
+          });
+    }
+    LOG.debug("Final set of compaction tmp files after removing active compactions: {}", matches);
+    return matches;
+  }
+
+  public static class DeleteStats {
+    public int success = 0;
+    public int failure = 0;
+    public int error = 0;
+  }
+
+  public static DeleteStats deleteTempFiles(ServerContext context, List<Path> filesToDelete)
+      throws InterruptedException {
+
+    final ExecutorService delSvc = Executors.newFixedThreadPool(8);
+    final List<Future<Boolean>> futures = new ArrayList<>(filesToDelete.size());
+    final DeleteStats stats = new DeleteStats();
+
+    filesToDelete.forEach(p -> {
+      futures.add(delSvc.submit(() -> context.getVolumeManager().delete(p)));
```

Comment:
Will this throw an exception when the files does not exists?  This util may find files that the system was in the process of renaming or deleting anyway, would be nice to only show errors when the file does not exists.

Reply:
I'm not sure if it would throw an exception, or just return false. The javadoc doesn't say. I guarded it with an exists(Path) in 9f4315a call just in case.